File: julia/Euler/src/bin/p0019.jl

# project euler: problem 19

module Prob0019

function solve_0019()
    common_year = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    leap_year = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

    # days per month (Jan 1901 - *Nov* 2000)
    days = repeat(vcat(repeat(common_year, 3), leap_year), 25)[1:end-1]

    # Jan 1, 1900 was Monday and assume that the day is the first day. (Monday is '1 mod 7 = 1')
    # And then, the year 1900 was a common year (365 days).
    # --> Jan 1, 1901 was Tuesday since (1 + 365) mod 7 = 2.
    #     Feb 1, 1901 was Firday since ((1 + 365) + 31) mod 7 = 5.
    #     ... and so on
    pushfirst!(days, 1 + 365)
    length(filter(x -> x % 7 == 0, accumulate(+, days)))
end

end #module

using .Prob0019: solve_0019
export solve_0019

File: go/internal/sol/p0019/p0019.go
package p0019

import (
	"slices"
	"strconv"
)

func compute() string {
	commonYear := []int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
	leapYear := []int{31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}

	// days per month (Jan 1901 - *Nov* 2000)
	days := slices.Repeat(slices.Concat(commonYear, commonYear, commonYear, leapYear), 25)
	days = days[:len(days)-1] // drop Dec 2000

	// insert number of days of the year 1900 (common year)
	days = slices.Insert(days, 0, 365)

	// accumulate
	for i := range len(days) - 1 {
		days[i+1] += days[i]
	}

	// The year 1900 was a comman year and January 1, 1900 was a Monday.
	// --> Monday: 0, Tuesday: 1, ..., Sunday: 6
	//     Jan 1, 1901 was Tuesday since 365 % 7 = 1.
	//     Feb 1, 1901 was Firday since (365 + 31) % 7 = 4.
	//     Mar 1, 1901 was Firday since (365 + 31 + 28) % 7 = 4.
	//     ... and so on
	var cnt int
	for x := range slices.Values(days) {
		// If it is Sunday, count it
		if x%7 == 6 {
			cnt++
		}
	}

	return strconv.FormatInt(int64(cnt), 10)
}

func Solve() string {
	return compute()
}
